### PR TITLE
[1265] Endpoint for updating sites

### PR DESCRIPTION
**Pull request by heidar**
### Context

Provider needs to be able to update site information.

### Changes proposed in this pull request

And update endpoint for sites that allows updating all attributes, except `code`.

### CODE REVIEW

**Comment by heidar** [comment]
Don't merge this yet, @tvararu is doing validations for the frontend that we can re-use in this PR. Should have that in tomorrow, otherwise we can merge this and get the validations in later. :)

**Comment by heidar** [comment]
> LGTM
> 
> We will need separate PRs:
> * add validations

We're doing validations in this one, we'll do a second PR for the API call.

File: app/models/site.rb [hunk 1]
```diff
@@ -21,4 +21,11 @@ class Site < ApplicationRecord
   include TouchProvider
 
   belongs_to :provider
+
+  validates :location_name, uniqueness: { scope: :provider }
```
Comment (defong) [inline app/models/site.rb]: 👍

File: app/models/concerns/postcode_normalize.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module PostcodeNormalize
+  extend ActiveSupport::Concern
+  included do
+    def postcode=(str)
+      if str
+        super UKPostcode.parse(str).to_s
+      else
+        super str
```
Comment (misaka) [inline app/models/concerns/postcode_normalize.rb]: MINOR: It looks like we're guarding against running `UKPostcode.parse(nil)` here, but it's not tested below. How important is it to guard against this and is it worth testing?

File: spec/factories/sites.rb [hunk 3]
```diff
@@ -24,7 +24,7 @@
     address2 { Faker::Address.community }
     address3 { Faker::Address.city }
     address4 { Faker::Address.state }
-    postcode { Faker::Address.postcode }
+    postcode { 'SW1A 2AA' }
```
Comment (misaka) [inline spec/factories/sites.rb]: Maybe set the Faker locale for this? Seems easy to do: https://github.com/stympy/faker/blob/master/lib/locales/README.md

File: app/models/concerns/postcode_normalize.rb (hunk 2 above)
Comment (misaka) [reply app/models/concerns/postcode_normalize.rb]: Explained by @tvararu: if we don't guard against `nil` here then many tests fail, so it's tested indirectly. LGTM.

Comment (heidar) [reply app/models/concerns/postcode_normalize.rb]: I added a unit test anyway

File: spec/factories/sites.rb (hunk 3 above)
Comment (tvararu) [reply spec/factories/sites.rb]: DOne

File: spec/rails_helper.rb [hunk 4]
```diff
@@ -12,6 +12,8 @@
 require 'rspec/rails'
 # Add additional requires below this line. Rails is not loaded until this point!
 
+Faker::Config.locale = 'en-GB'
```
Comment (misaka) [inline spec/rails_helper.rb]: 💯

File: spec/requests/api/v2/providers/sites_spec.rb [hunk 5]
```diff
@@ -133,4 +133,217 @@
...
+        )
+      end
+
+      it 'updates the location name of the site' do
+        expect { subject }.to change { site1.reload.location_name }
+          .from(site1.location_name)
+          .to(location_name)
+      end
+
+      it 'updates the code of the site' do
```
Comment (misaka) [inline spec/requests/api/v2/providers/sites_spec.rb]: ITYM: DOES NOT update the code of the site :)

File: spec/requests/api/v2/providers/sites_spec.rb [hunk 6]
```diff
@@ -133,4 +133,217 @@
...
+              expect(response.body).to include('Invalid postcode')
+              expect(response.body).to include("Postcode can't be blank")
+            end
+
+            it 'checks the postcode is present' do
+              expect(response.body).to include('Invalid postcode')
+              expect(response.body).to include('Postcode not recognised as a UK postcode')
+            end
+
+            it 'checks the region_code is present' do
```
Comment (misaka) [inline spec/requests/api/v2/providers/sites_spec.rb]: Maybe use `xit` to flag this as a pending test?

**Pull request merged**